odoo/tools/misc.py

```diff
@@ -1629,6 +1633,69 @@ def hmac(env, scope, message, hash_function=hashlib.sha256):
     ).hexdigest()
 
 
+def hash_sign(env, scope, message_values, expiration=None, expiration_hours=None, hash_function=hashlib.sha256):
+    """ Generate an urlsafe payload signed with the HMAC signature for an iterable set of data.
+
+    :param env: sudo environment to use for retrieving config parameter
+    :param scope: scope of the authentication, to have different signature for the same
+        message in different usage
+    :param message_values: values to be encoded inside the payload
+    :param expiration: optional, a datetime or timedelta
+    :param expiration_hours: optional, a int representing a number of hours before expiration. Cannot be set at the same time as expiration
+    :return: the payload that can be used as a token
+    """
+    assert not (expiration and expiration_hours)
+    assert message_values is not None
+
+    if expiration_hours:
+        expiration = datetime.datetime.now() + datetime.timedelta(hours=expiration_hours)
+    else:
+        if isinstance(expiration, datetime.timedelta):
+            expiration = datetime.now() + expiration
+    expiration_timestamp = 0 if not expiration else int(expiration.timestamp())
+
+    # we need to encode in base64 individually before adding the separator to avoid any collision between the content and the separator.
+    message_strings = json.dumps(message_values)
+    message_strings_encoded = message_strings.encode()
+    b64_message_list = base64.urlsafe_b64encode(message_strings_encoded)
+    b64_message = b64_message_list.decode().rstrip('=')
+
+    hash_value = hmac(env, scope, f'{b64_message}:{expiration_timestamp}')
+
+    b64_expiration = base64.urlsafe_b64encode(expiration_timestamp.to_bytes(math.ceil(expiration_timestamp.bit_length() / 8), 'little')).decode().rstrip("=")
+    b64_hash = base64.urlsafe_b64encode(bytes.fromhex(hash_value)).decode().rstrip("=")
+
+    return f'1~{b64_message}~{b64_expiration}~{b64_hash}'
+
+
+def verify_hash_signed(env, scope, payload, hash_function=hashlib.sha256):
+    """ Verify and extract data from a given urlsafe  payload generated with hash_sign()
+
+    :param env: sudo environment to use for retrieving config parameter
+    :param scope: scope of the authentication, to have different signature for the same
+        message in different usage
+    :param payload: the token to verify
+    :return: The payload_values if the check was successful, None otherwise.
+    """
+    version, b64_message, b64expiration_value, b64hash_value = payload.split("~")
+
+    if version != '1':
+        raise UserError('Unknown token version')
```

Comment:
either we raise a ValueError, either we translate the error (but I vote for ValueError)